packages/core/src/pipeline/config/validation/imageProviderBeforeType.validator.ts

```diff
@@ -0,0 +1,30 @@
+import { IStageOrTriggerValidator, IValidatorConfig, PipelineConfigValidator } from './PipelineConfigValidator';
+import { ICluster, IPipeline, IStage, IStageOrTriggerTypeConfig } from '../../../domain';
+
+export interface IImageProviderBeforeTypeValidationConfig extends IValidatorConfig {
+  triggerTypes?: string[];
+  message: string;
+}
+
+export class ImageProviderBeforeTypeValidator implements IStageOrTriggerValidator {
+  public validate( 
+    pipeline: IPipeline,
+    stage: IStage,
+    validator: IImageProviderBeforeTypeValidationConfig,
+    _config: IStageOrTriggerTypeConfig,
+  ): string {
+
+    const triggersToTest = pipeline.triggers || [];
+    const hasImageProvider = Boolean(triggersToTest.length) && triggersToTest.every((trigger) => (validator.triggerTypes || []).includes(trigger.type));
+
+    const hasCustomImage = (stage.clusters || []).every((cluster: ICluster) => cluster.imageId && cluster.imageId !== '${trigger.properties.imageName}');
+
+    if (!hasImageProvider && !hasCustomImage) {
+      return `${validator.message} Add a trigger or manually input the imageId. Suggested triggers: ${(validator.triggerTypes || []).join(', ')}`;
```

Comment:
I'm not familiar with the pipeline validator design but curious to know why the validation and the message is split across two places?

Reply:
They are split up so that the message can be configured based on the stage that is used. A good example is the `stageBeforeType` validator. Both the canary and deploy stage use this validator, but have a different warning message if the validation fails.

The messages are mostly similar. If there was ever a validator refactor, then I think the message should be an optional override.